### PR TITLE
Example on consistency training for vision models

### PR DESCRIPTION
**Pull request by sayakpaul**
This example demonstrates _consistency training_ which is one of the major elements of works like FixMatch, UDA, and Noisy Student Training. It not only yields better performing models having higher topline metrics like Top-1 Accuracy but also provides enhanced robustness to common corruptions and perturbations.

Besides, this example could also be treated as a template for developing semi-supervised or weakly supervised classification methods.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Not sure what is happening with the black formatting since I did run `black consistency_training.py` before committing it.

**Comment by fchollet** [comment]
> Not sure what is happening with the black formatting since I did run black consistency_training.py before committing it.

Maybe a discrepancy between `black` versions?

**Comment by sayakpaul** [comment]
> Maybe a discrepancy between `black` versions?

That seems to be the case. After reformatting now when I am trying to make the final push it says everything is up-to-date:

image

**Comment by 8bitmp3** [comment]
This is an excellent example and thanks for bringing up the challenges of non-i.i.d. data and what this means for computer vision tasks.

I'd suggest to have the introductory section:

- Explain what the example does in the first sentence (why should users care about this awesome example, since they are busy)
- Mention what can cause the distribution shifts (such as random noise, contrast change, and blurring)
- Mention that normally we assume i.i.d. but these models can have performance degradation (this is from the "A Fourier Perspective on Model Robustness in Computer Vision" paper you mentioned)
- Trim words without altering the meaning (e.g. "conventional... standard..." in one sentence)
- Mention the name of the great paper ("A Fourier Perspective on Model Robustness in Computer Vision" vs "this literature", so that the users don't have to lick to know what it is)
- Other stuff related to syntax and style

```
This example demonstrates how to achieve robustness against data distribution
shifts in computer vision with the help of the
[RandAugment](https://arxiv.org/abs/1909.13719) data augmentation method.

Deep learning models excel in many image recognition tasks when the data is
independent and identically distributed (i.i.d.). However, they can suffer from
performance degradation caused by subtle distribution shifts in the input data 
(such as random noise, contrast change, and blurring), as discussed in
[A Fourier Perspective on Model Robustness in Computer Vision](https://arxiv.org/pdf/1906.08988.pdf)).
In addition, model training procedures, such as standard image classification
training workflows) don't typically enable a model to generalize well.

In this example, we will be training an image classification model enforcing a
sense of *consistency* inside it through the following steps:

- Train a standard image classification model.
- Train an equal or larger model on a noisy version of the dataset augmented
using [RandAugment](https://arxiv.org/abs/1909.13719). 
  - To do this, we begin with obtaining predictions of the previous model on the
  clean images of the dataset. 
  - Then, we use these predictions and train the second model to match these
predictions on the noisy variant of the same images. This is identical to the
workflow in the [*Knowledge Distillation*](https://keras.io/examples/vision/knowledge_distillation/)
example; but, because the student model is equal or larger in size, this process
is also referred to as *self-training*. 
```

**Comment by 8bitmp3** [comment]
Regarding the title, because "consistency" may mean different things to different users in different domains, I'd suggest the following to be more precise and capture the right audience (The current title is very vague IMHO)

```suggestion
Title: Robustness against data distribution shifts in supervised learning in computer vision
```

"Robustness against data distribution shifts" is how the authors of "a Fourier Perspective on Model Robustness in Computer Vision" paper would probably describe it.

**Comment by 8bitmp3** [comment]
Alternatively, you could alter the description to  with "supervised learning" -> "supervised learning in computer vision" (for precision)

```
Description: Robustness against data distribution shifts in supervised learning in computer vision
```

While leaving the title shorter, such as:

```
Title: Robustness against data distribution shifts
```

**Comment by sayakpaul** [comment]
I would keep the description as you suggested but for the title, I would prefer the current one. Here's why:

Although the example focuses on using consistency training to enhance the robustness of models to common corruptions this example also serves as a template for weakly supervised learning and other forms of consistency regularization. Maybe mentioning this in the text would be more justifiable.

File: examples/vision/consistency_training.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+ensembling* and when the student model is forced to mimic that, it leads to better
+performance. The original work uses [Stochastic Depth](https://arxiv.org/abs/1603.09382)
+and [Dropout](https://jmlr.org/papers/v15/srivastava14a.html) to bring in the ensembling
+part but for this example, we will use [Stochastic Weight Averaging](https://arxiv.org/abs/1803.05407)
+(SWA) which also resembles geometric ensembling.
+"""
+
+# Define the callbacks.
+reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(patience=3)
+es = tf.keras.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
```
Comment (fchollet) [inline examples/vision/consistency_training.py:220]: Use fully spelled out variable names

File: examples/vision/consistency_training.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+# Initialize SWA from tf-hub.
+SWA = tfa.optimizers.SWA
```
Comment (fchollet) [inline examples/vision/consistency_training.py:223]: How are you using this?

File: examples/vision/consistency_training.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+        # Return a dict of performance
+        results = {m.name: m.result() for m in self.metrics}
+        return results
+
+
+"""
+The only difference in this implementation is the way loss is being calculated. Instead
+of weighted the distillation loss and student loss differently we are taking their
+average following Noisy Student Training.
```
Comment (fchollet) [inline examples/vision/consistency_training.py:329]: Here, it would be worth highlighting the specific line in question

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This is a great example.

File: examples/vision/consistency_training.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+"""shell
+pip install -q tf-models-official tensorflow-addons
+"""
+
+"""
+## Imports and setup
+"""
+
+from official.vision.image_classification.augment import RandAugment
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/consistency_training.py:55]: Yes. It's been some time. I got to know about it via the following tutorials:

* https://www.tensorflow.org/tutorials/text/classify_text_with_bert
* https://www.tensorflow.org/tutorials/text/solve_glue_tasks_using_bert_on_tpu

File: examples/vision/consistency_training.py (hunk 2 above)
Comment (sayakpaul) [reply examples/vision/consistency_training.py:223]: Yes. I had forgotten to use it to wrap the optimizer. It's been fixed now and I have also added a comment above its usage so that it won't go unnoticed.

File: examples/vision/consistency_training.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/vision/consistency_training.py:54]: I've escalated this, hopefully it will get resolved.

**Comment by sayakpaul** [comment]
@fchollet would suggest opening a separate PR since the Black CI is failing?

**Comment by fchollet** [comment]
> @fchollet would suggest opening a separate PR since the Black CI is failing?

We can merge first and fix it afterwards. We should also look into making the failure message more informative, as a follow up.

**Review by fchollet — APPROVED** [review]
Looking great -- thanks for the update! We can add the generated files now, then merge, then manually resolve the style issue.

**Comment by sayakpaul** [comment]
Thanks, @fchollet. Should be all set now.

**Review by fchollet — APPROVED** [review]
Excellent -- thank you for the great contribution!
